File: app/src/main/java/com/blanke/mdwechat/WechatHook.java
package com.blanke.mdwechat;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.res.XModuleResources;

import de.robv.android.xposed.IXposedHookInitPackageResources;
import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.IXposedHookZygoteInit;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_InitPackageResources;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import static com.blanke.mdwechat.ui.LogUtil.log;

public class WechatHook extends XC_MethodHook
        implements IXposedHookZygoteInit, IXposedHookLoadPackage, IXposedHookInitPackageResources {

    private static String MODULE_PATH = null;

    @Override
    public void initZygote(StartupParam startupParam) throws Throwable {
        MODULE_PATH = startupParam.modulePath;
        WeChatHelper.initPrefs();
    }

    @Override
    public void handleLoadPackage(final XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        if (!lpparam.packageName.equals(Common.WECHAT_PACKAGENAME)) {
            return;
        }
        Context context = (Context) XposedHelpers.callMethod(
                XposedHelpers.callStaticMethod(XposedHelpers.findClass("android.app.ActivityThread", null),
                        "currentActivityThread"), "getSystemContext");
        PackageInfo wechatPackageInfo = context.getPackageManager().getPackageInfo(Common.WECHAT_PACKAGENAME, 0);
        String versionName = wechatPackageInfo.versionName;
        log("wechat version=" + versionName
                + ",processName=" + lpparam.processName
                + ",MDWechat version=" + BuildConfig.VERSION_NAME);
        WeChatHelper.init(versionName, lpparam);
    }

    private void hookSettings(XC_LoadPackage.LoadPackageParam lpparam, final String versionName, final boolean isSupport) {
        XposedHelpers.findAndHookMethod("com.blanke.mdwechat.settings.SettingsFragment",
                lpparam.classLoader, "getWeChatVersion", new XC_MethodHook() {
                    @Override
                    protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                        param.setResult(versionName);
                    }
                });
        XposedHelpers.findAndHookMethod("com.blanke.mdwechat.settings.SettingsFragment",
                lpparam.classLoader, "isSupportWechat", new XC_MethodHook() {
                    @Override
                    protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                        param.setResult(isSupport);
                    }
                });

    }

    @Override
    public void handleInitPackageResources(XC_InitPackageResources.InitPackageResourcesParam resparam) throws Throwable {
        if (!resparam.packageName.equals(Common.WECHAT_PACKAGENAME)) {
            return;
        }
//        log("handleInitPackageResources");
        XModuleResources modRes = XModuleResources.createInstance(MODULE_PATH, resparam.res);
//        WeChatHelper.WCId.ic_add = resparam.res.addResource(modRes, R.drawable.ic_add);


//        resparam.res.setReplacement(Common.WECHAT_PACKAGENAME,
//                "color", "z", Color.RED);
//        resparam.res.setReplacement(Common.WECHAT_PACKAGENAME,
//                "drawable", WeChatHelper.WCDrawable.Conference_ListView_Item_Background,
//                modRes.fwd(R.drawable.selector_item));
    }
}

File: app/src/main/java/com/blanke/mdwechat/WechatHook.kt
package com.blanke.mdwechat

import com.blanke.mdwechat.config.HookConfig
import com.blanke.mdwechat.config.WxVersionConfig
import com.blanke.mdwechat.hookers.*
import com.blanke.mdwechat.hookers.base.HookerProvider
import com.blanke.mdwechat.util.LogUtil.log
import de.robv.android.xposed.IXposedHookLoadPackage
import de.robv.android.xposed.callbacks.XC_LoadPackage

class WechatHook : IXposedHookLoadPackage {

    @Throws(Throwable::class)
    override fun handleLoadPackage(lpparam: XC_LoadPackage.LoadPackageParam) {
        try {
            if (!(lpparam.packageName.contains("com.tencent") && lpparam.packageName.contains("mm")))
                return
            // 暂时不 hook 小程序
            if (lpparam.processName.contains(":")) {
                return
            }
            WeChatHelper.initPrefs()
            if (!HookConfig.is_hook_switch) {
                log("模块总开关已关闭")
                return
            }
            log("模块加载成功")
            val hookers = mutableListOf(
                    LauncherUIHooker,
                    ActionBarHooker,
                    StatusBarHooker,
                    AvatarHooker,
                    ListViewHooker,
                    ConversationHooker,
                    ContactHooker,
                    DiscoverHooker,
                    SettingsHooker,
                    SchemeHooker,
                    LogHooker
            )
            if (BuildConfig.DEBUG) {
                hookers.add(0, DebugHooker)
            }
            hookMain(lpparam, hookers)
        } catch (e: Throwable) {
            log(e)
        }
    }

    private fun hookMain(lpparam: XC_LoadPackage.LoadPackageParam, plugins: List<HookerProvider>) {
        WechatGlobal.init(lpparam)
        try {
            WechatGlobal.wxVersionConfig = WxVersionConfig.loadConfig(WechatGlobal.wxVersion!!.toString())
        } catch (e: Exception) {
            log("${WechatGlobal.wxVersion} 配置文件不存在或解析失败")
            return
        }
        log("wechat version=" + WechatGlobal.wxVersion
                + ",processName=" + lpparam.processName
                + ",MDWechat version=" + BuildConfig.VERSION_NAME)
        plugins.forEach { provider ->
            provider.provideStaticHookers()?.forEach { hooker ->
                if (!hooker.hasHooked) {
                    hooker.hook()
                    hooker.hasHooked = true
                }
            }
        }
    }
}

